File: Indicators/Sma/Sma.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Skender.Stock.Indicators
{
    public static partial class Indicator
    {
        // SIMPLE MOVING AVERAGE
        public static IEnumerable<SmaResult> GetSma(IEnumerable<Quote> history, int lookbackPeriod)
        {

            // clean quotes
            Cleaners.PrepareHistory(history);

            // check parameters
            ValidateSma(history, lookbackPeriod);

            // initialize
            List<SmaResult> results = new List<SmaResult>();

            // roll through history
            foreach (Quote h in history)
            {

                SmaResult result = new SmaResult
                {
                    Index = (int)h.Index,
                    Date = h.Date
                };

                if (h.Index >= lookbackPeriod)
                {
                    List<Quote> period = history
                        .Where(x => x.Index <= h.Index && x.Index > (h.Index - lookbackPeriod))
                        .ToList();

                    // simple moving average
                    result.Sma = period
                        .Select(x => x.Close)
                        .Average();

                    // mean absolute deviation
                    result.Mad = period
                        .Select(x => Math.Abs(x.Close - (decimal)result.Sma))
                        .Average();

                    // mean squared error
                    result.Mse = period
                        .Select(x => (x.Close - (decimal)result.Sma) * (x.Close - (decimal)result.Sma))
                        .Average();

                    // mean absolute percent error
                    result.Mape = period
                        .Where(x => x.Close != 0)
                        .Select(x => Math.Abs(x.Close - (decimal)result.Sma) / x.Close)
                        .Average();
                }

                results.Add(result);
            }

            return results;
        }


        private static void ValidateSma(IEnumerable<Quote> history, int lookbackPeriod)
        {

            // check parameters
            if (lookbackPeriod <= 0)
            {
                throw new BadParameterException("Lookback period must be greater than 0 for SMA.");
            }

            // check history
            int qtyHistory = history.Count();
            int minHistory = lookbackPeriod;
            if (qtyHistory < minHistory)
            {
                throw new BadHistoryException("Insufficient history provided for SMA.  " +
                        string.Format(englishCulture,
                        "You provided {0} periods of history when at least {1} is required.",
                        qtyHistory, minHistory));
            }

        }
    }

}

File: wraps/python/src/indicators/SMA.py
from Skender.Stock.Indicators import Indicator
from cstypes import List, to_pydecimal
from .Qoute import Qoute

def get_sma(history, lookbackPeriod: int):
    sma_list = Indicator.GetSma[Qoute](List(Qoute, history), lookbackPeriod)
    sma_list = [ SmaResult(r) for r in sma_list ]

    return sma_list

def get_sma_extended(history, lookbackPeriod: int):
    sma_extended_list = Indicator.GetSmaExtended[Qoute](List(Qoute, history), lookbackPeriod)
    sma_extended_list = [ SmaExtendedResult(r) for r in sma_extended_list ]

    return sma_extended_list

class SmaResult:
    def __init__(self,sma_result):
        self.Date = sma_result.Date
        self.Sma = to_pydecimal(sma_result.Sma)

    def __str__(self):
        return str(self.Date) + ": " + str(self.Sma)

class SmaExtendedResult(SmaResult):
    def __init__(self, sma_extended_result):
        super().__init__(sma_extended_result)
        self.Mad = to_pydecimal(sma_extended_result.Mad)
        self.Mse = to_pydecimal(sma_extended_result.Mse)
        self.Mape = to_pydecimal(sma_extended_result.Mape)

    def __str__(self):
        return str(self.Date) + ": " + str(self.Sma) + "\t"\
             + str(self.Mad) + "\t" + str(self.Mse) + "\t"\
             + str(self.Mape)
    

